presto-native-execution/presto_cpp/main/SystemConnector.h

```diff
@@ -0,0 +1,233 @@
+      std::string schemaName,
+      std::string tableName);
+
+  ~SystemTableHandle() override {}
+
+  std::string toString() const override;
+
+  const std::string& schemaName() {
+    return schemaName_;
+  }
+
+  const std::string& tableName() {
+    return tableName_;
+  }
+
+  const velox::RowTypePtr& taskSchema();
+
+ private:
+  const std::string schemaName_;
+  const std::string tableName_;
+};
+
+class SystemDataSource : public velox::connector::DataSource {
+ public:
+  SystemDataSource(
+      const velox::RowTypePtr& outputType,
+      const std::shared_ptr<velox::connector::ConnectorTableHandle>&
+          tableHandle,
+      const std::unordered_map<
+          std::string,
+          std::shared_ptr<velox::connector::ColumnHandle>>& columnHandles,
+      const TaskManager* taskManager,
+      velox::memory::MemoryPool* pool);
+
+  void addSplit(
+      std::shared_ptr<velox::connector::ConnectorSplit> split) override;
+
+  void addDynamicFilter(
+      velox::column_index_t /*outputChannel*/,
+      const std::shared_ptr<velox::common::Filter>& /*filter*/) override {
+    VELOX_NYI("Dynamic filters not supported by SystemConnector.");
+  }
+
+  std::optional<velox::RowVectorPtr> next(
+      uint64_t size,
+      velox::ContinueFuture& future) override;
+
+  uint64_t getCompletedRows() override {
+    return completedRows_;
+  }
+
+  uint64_t getCompletedBytes() override {
+    return completedBytes_;
+  }
+
+  std::unordered_map<std::string, velox::RuntimeCounter> runtimeStats()
+      override {
+    // TODO: Which stats do we want to expose here?
+    return {};
+  }
+
+ private:
+  enum class TaskColumnEnum {
+    // Note: These values are in the same order as SystemTableHandle schema.
+    kNodeId = 0,
+    kTaskId,
+    kStageExecutionId,
+    kStageId,
+    kQueryId,
+    kState,
+    kSplits,
+    kQueuedSplits,
+    kRunningSplits,
+    kCompletedSplits,
+    kSplitScheduledTimeMs,
+    kSplitCpuTimeMs,
+    kSplitBlockedTimeMs,
+    kRawInputBytes,
+    kRawInputRows,
+    kProcessedInputBytes,
+    kProcessedInputRows,
+    kOutputBytes,
+    kOutputRows,
+    kPhysicalWrittenBytes,
+    kCreated,
+    kStart,
+    kLastHeartBeat,
+    kEnd,
+  };
+
+  velox::RowVectorPtr getTaskResults();
+
+  // Mapping between output columns and their indices (column_index_t)
+  // corresponding to the taskInfo fields for them.
+  std::vector<velox::column_index_t> outputColumnMappings_;
+  velox::RowTypePtr outputType_;
+
+  const TaskManager* taskManager_;
+  velox::memory::MemoryPool* pool_;
+
+  std::shared_ptr<SystemSplit> currentSplit_;
+
+  size_t completedRows_{0};
+  size_t completedBytes_{0};
+};
+
+class SystemConnector : public velox::connector::Connector {
+ public:
+  SystemConnector(
+      const std::string& id,
+      std::shared_ptr<const velox::Config> /*properties*/,
+      folly::Executor* /*executor*/)
+      : Connector(id) {}
+
+  void setTaskManager(const TaskManager* taskManager) {
```

Comment:
Having a setter makes it prone to a `null` pointer related bugs (if somebody creates a `SystemConnector` but forgets to set a TaskManager)

I wonder if we even need the `SystemConnectorFactory`? Since the connector is manually registered I wonder whether it makes sense to simply take the `TaskManager` in the constructor and create it without using a factory?

Reply:
Good point. This code can be simplified and made more robust after removing SystemConnectorFactory. And there isn't a need for it really.